atc/configvalidate/validate.go

```diff
@@ -679,3 +683,42 @@ func compositeErr(errorMessages []string) error {
 
 	return errors.New(strings.Join(errorMessages, "\n"))
 }
+
+func validateCredentialManagers(c Config) error {
+	for _, cm := range c.CredentialManagers {
+		factory := creds.ManagerFactories()[cm.Type]
+		if factory == nil {
+			return fmt.Errorf("unknown credential manager type: %s", cm.Type)
+		}
+
+		// TODO: this check should eventually be removed once all credential managers
+		// are supported in pipeline. - @evanchaoli
+		switch cm.Type {
+		case "vault":
+		default:
+			return fmt.Errorf("credential manager type %s is not supported in pipeline yet", cm.Type)
+		}
+
+		manager := factory.NewInstance()
+
+		var (
+			b   []byte
+			err error
+		)
+		b, err = json.Marshal(cm.Config)
+		if err != nil {
+			return err
+		}
+
+		err = json.Unmarshal(b, &manager)
```

Comment:
Let me investigate a little bit. Fortunately this doesn't affect design of this feature.

Reply:
You can have a look at the skymarshal bits [here](https://github.com/concourse/concourse/blob/master/skymarshal/skycmd/flags.go#L77).

It's not pretty, but it does reduce the amount of boilerplate you need to write when adding new auth connectors.

Reply:
@pivotal-jwinters and @vito To be honest, I don't clearly the point of `shymarshal`. I check code that Joshua pointed out, looks like the key is to use `github.com/mitchellh/mapstructure`. But with that, we will still need to use `mapstructure`, because: 1) we need to squash struct because `TLS` and `AuthConfig` have second level data, 2) parameter name is not same to struct field name, like "ca-path" vs "CAPath".

For `spf13/viper`, I guess I cannot just use it for `var_sources`, so I cannot use it for this PR.

Did I check the right thing you want to me check?

Reply:
@evanchaoli Let's just skip over this for now, just do what's easiest and low-ish maintenance and we can re-assess things later on. I think the number of credential managers to support will be lower than the number of auth providers so this hopefully won't be as painful. :)